File: tools/make_bart_sprites.ts
#!/usr/bin/env node
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import {
  cloneImage,
  createImage,
  crop,
  getPixel,
  parseHex,
  readPng,
  resizeNearest,
  setPixel,
  type PixelImage,
  type Rgba,
  writePng,
} from './lib/pixel';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const repoRoot = path.resolve(__dirname, '..');

const SOURCE = path.join(repoRoot, 'public/assets/bart_source.png');
const OUTPUT_DIR = path.join(repoRoot, 'public/assets/sprites');

const OUTPUTS = [
  { file: 'bart_head_32.png', size: 32, sample: 10 },
  { file: 'bart_head_48.png', size: 48, sample: 14 },
  { file: 'bart_head_64.png', size: 64, sample: 18 },
  { file: 'bart_portrait_96.png', size: 96, sample: 24 },
] as const;

const WHITE_THRESHOLD = 244;
const MARGIN_RATIO = 0.08;
const OUTLINE: Rgba = [22, 22, 24, 220];

const BART_PALETTE: Rgba[] = [
  parseHex('#101010'),
  parseHex('#1D1D1D'),
  parseHex('#2B2824'),
  parseHex('#412F2A'),
  parseHex('#5D4438'),
  parseHex('#754B3D'),
  parseHex('#8D5C4A'),
  parseHex('#A66E54'),
  parseHex('#BF8263'),
  parseHex('#D89A76'),
  parseHex('#E9B48E'),
  parseHex('#F2C98E'),
  parseHex('#E6D05A'),
  parseHex('#F1DD77'),
  parseHex('#F4E3A1'),
  parseHex('#9D2C2C'),
  parseHex('#B63A3A'),
  parseHex('#CF5151'),
  parseHex('#E36863'),
  parseHex('#F3F3F3'),
];

function maskWhiteBackground(source: PixelImage): PixelImage {
  const masked = cloneImage(source);
  for (let y = 0; y < masked.height; y += 1) {
    for (let x = 0; x < masked.width; x += 1) {
      const [r, g, b, a] = getPixel(masked, x, y);
      if (r >= WHITE_THRESHOLD && g >= WHITE_THRESHOLD && b >= WHITE_THRESHOLD) {
        setPixel(masked, x, y, [r, g, b, 0]);
      } else {
        setPixel(masked, x, y, [r, g, b, a]);
      }
    }
  }
  return masked;
}

function alphaBounds(image: PixelImage): { minX: number; minY: number; maxX: number; maxY: number } {
  let minX = image.width;
  let minY = image.height;
  let maxX = -1;
  let maxY = -1;

  for (let y = 0; y < image.height; y += 1) {
    for (let x = 0; x < image.width; x += 1) {
      const [, , , alpha] = getPixel(image, x, y);
      if (alpha === 0) {
        continue;
      }
      minX = Math.min(minX, x);
      minY = Math.min(minY, y);
      maxX = Math.max(maxX, x);
      maxY = Math.max(maxY, y);
    }
  }

  if (maxX < minX || maxY < minY) {
    return { minX: 0, minY: 0, maxX: image.width - 1, maxY: image.height - 1 };
  }

  return { minX, minY, maxX, maxY };
}

function cropSquare(image: PixelImage): PixelImage {
  const bounds = alphaBounds(image);
  const width = bounds.maxX - bounds.minX + 1;
  const height = bounds.maxY - bounds.minY + 1;
  const side = Math.ceil(Math.max(width, height) * (1 + MARGIN_RATIO * 2));
  const centerX = bounds.minX + Math.floor(width / 2);
  const centerY = bounds.minY + Math.floor(height / 2);

  const left = Math.max(0, Math.min(image.width - side, centerX - Math.floor(side / 2)));
  const top = Math.max(0, Math.min(image.height - side, centerY - Math.floor(side / 2)));

  return crop(image, left, top, side, side);
}

function nearestPaletteColor(r: number, g: number, b: number): Rgba {
  let best = BART_PALETTE[0];
  let bestDistance = Number.POSITIVE_INFINITY;
  for (const candidate of BART_PALETTE) {
    const dr = r - candidate[0];
    const dg = g - candidate[1];
    const db = b - candidate[2];
    const distance = dr * dr + dg * dg + db * db;
    if (distance < bestDistance) {
      best = candidate;
      bestDistance = distance;
    }
  }
  return best;
}

function quantizePalette(image: PixelImage): PixelImage {
  const quantized = createImage(image.width, image.height, [0, 0, 0, 0]);
  for (let y = 0; y < image.height; y += 1) {
    for (let x = 0; x < image.width; x += 1) {
      const [r, g, b, a] = getPixel(image, x, y);
      if (a === 0) {
        continue;
      }
      const nearest = nearestPaletteColor(r, g, b);
      setPixel(quantized, x, y, [nearest[0], nearest[1], nearest[2], 255]);
    }
  }
  return quantized;
}

function addSubtleOutline(image: PixelImage): PixelImage {
  const outlined = cloneImage(image);
  const snapshot = cloneImage(image);

  for (let y = 0; y < image.height; y += 1) {
    for (let x = 0; x < image.width; x += 1) {
      const [, , , alpha] = getPixel(snapshot, x, y);
      if (alpha > 0) {
        continue;
      }
      const neighbors = [
        getPixel(snapshot, x - 1, y),
        getPixel(snapshot, x + 1, y),
        getPixel(snapshot, x, y - 1),
        getPixel(snapshot, x, y + 1),
      ];
      if (neighbors.some((neighbor) => neighbor[3] > 0)) {
        setPixel(outlined, x, y, OUTLINE);
      }
    }
  }

  return outlined;
}

function main(): number {
  let source: PixelImage;
  try {
    source = readPng(SOURCE);
  } catch {
    console.error(`ERROR: missing or invalid source image: ${SOURCE}`);
    return 1;
  }

  const masked = maskWhiteBackground(source);
  const cropped = cropSquare(masked);

  for (const output of OUTPUTS) {
    const downscaled = resizeNearest(cropped, output.sample, output.sample);
    const quantized = quantizePalette(downscaled);
    const upscaled = resizeNearest(quantized, output.size, output.size);
    const outlined = addSubtleOutline(upscaled);

    const outputPath = path.join(OUTPUT_DIR, output.file);
    writePng(outputPath, outlined);
    console.log(`Wrote ${path.relative(repoRoot, outputPath)} (${output.size}x${output.size})`);
  }

  return 0;
}

process.exitCode = main();

File: tools/make_bart_sprites.py
#!/usr/bin/env python3
"""Generate pixelated Bart avatar sprites from public/assets/target_look.png."""

from __future__ import annotations

from pathlib import Path
import sys

try:
    from PIL import Image, ImageChops, ImageFilter
except Exception:  # pragma: no cover
    print('ERROR: Pillow is required. Run: python3 -m pip install -r tools/requirements.txt')
    raise SystemExit(1)

MARGIN_RATIO = 0.08
COLORS = 24
OUTLINE_RGBA = (18, 18, 18, 255)

OUTPUTS = {
    'bart_head_32.png': (32, 8),
    'bart_head_48.png': (48, 12),
    'bart_head_64.png': (64, 16),
    'bart_portrait_96.png': (96, 32),
}


def center_crop_with_margin(image: Image.Image) -> Image.Image:
    width, height = image.size
    side = min(width, height)
    side = max(1, int(round(side * (1.0 - MARGIN_RATIO))))
    left = (width - side) // 2
    top = (height - side) // 2
    return image.crop((left, top, left + side, top + side))


def pixelate(image: Image.Image, target: int, working: int) -> Image.Image:
    small = image.resize((working, working), Image.Resampling.BOX)
    quant = small.convert('RGB').quantize(
        colors=COLORS,
        method=Image.Quantize.MEDIANCUT,
        dither=Image.Dither.FLOYDSTEINBERG,
    ).convert('RGBA')
    return quant.resize((target, target), Image.Resampling.NEAREST)


def add_outline(image: Image.Image) -> Image.Image:
    rgba = image.convert('RGBA')
    alpha = rgba.split()[3]
    expanded = alpha.filter(ImageFilter.MaxFilter(3))
    outline_mask = ImageChops.subtract(expanded, alpha)
    out = rgba.copy()
    out.paste(OUTLINE_RGBA, (0, 0), outline_mask)
    return out


def main() -> int:
    repo_root = Path(__file__).resolve().parents[1]
    source_path = repo_root / 'public' / 'assets' / 'target_look.png'
    sprites_dir = repo_root / 'public' / 'assets' / 'sprites'

    if not source_path.exists():
        print(f'ERROR: missing source portrait: {source_path}')
        return 1

    sprites_dir.mkdir(parents=True, exist_ok=True)

    source = Image.open(source_path).convert('RGBA')
    cropped = center_crop_with_margin(source)

    for filename, (target_size, work_size) in OUTPUTS.items():
        img = pixelate(cropped, target_size, work_size)
        img = add_outline(img)
        out_path = sprites_dir / filename
        img.save(out_path, format='PNG', optimize=True)
        print(f'Wrote {out_path.relative_to(repo_root)} ({target_size}x{target_size})')

    return 0


if __name__ == '__main__':
    raise SystemExit(main())
